syntax = "proto3";

// import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";
import "google/protobuf/any.proto";
import "google/protobuf/wrappers.proto";

package doodle;

message file_info_server{
  string path = 1;
  bool is_folder = 2;
  bool exist = 3;
  google.protobuf.Timestamp update_time = 4;
  uint64 size = 5;
  google.protobuf.StringValue hash = 6;
}

message file_stream_server{
  file_info_server info = 1;
  google.protobuf.Any data = 2;
}

message file_info_move_server{
  file_info_server source = 1;
  file_info_server target = 2;
}

service file_system_server{
  rpc get_info(file_info_server) returns (file_info_server) {}
  rpc get_hash(file_info_server) returns (file_info_server) {}

  rpc is_exist(file_info_server) returns (file_info_server) {}
  rpc get_size(file_info_server) returns (file_info_server) {}
  rpc is_folder(file_info_server) returns (file_info_server) {}
  rpc get_timestamp(file_info_server) returns (file_info_server) {}
  rpc get_list(file_info_server) returns (stream file_info_server) {}

  rpc download(file_info_server) returns (stream file_stream_server) {}
  rpc upload(stream file_stream_server) returns (file_info_server) {}
  rpc move(file_info_move_server) returns (file_info_server) {}
}
